provider/kubernetes/kubernetes_test.go

```diff
@@ -797,6 +798,166 @@ func TestIngressAnnotations(t *testing.T) {
+			),
+		),
+		buildIngress(
+			iNamespace("testing"),
+			iAnnotation(annotationKubernetesIngressClass, traefikDefaultAnnotationValue+"-other"),
+			iRules(
+				iRule(
+					iHost("herp"),
+					iPaths(onePath(iPath("/derp"), iBackend("service2", intstr.FromInt(80))))),
+			),
+		),
+	}
+
+	services := []*v1.Service{
+		buildService(
+			sName("service1"),
+			sNamespace("testing"),
+			sUID("1"),
+			sSpec(
+				clusterIP("10.0.0.1"),
+				sType("ExternalName"),
+				sExternalName("example.com"),
+				sPorts(sPort(80, "http"))),
+		),
+		buildService(
+			sName("service2"),
+			sNamespace("testing"),
+			sUID("2"),
+			sSpec(
+				clusterIP("10.0.0.2"),
+				sPorts(sPort(802, ""))),
+		),
+	}
+
+	watchChan := make(chan interface{})
+	client := clientMock{
+		ingresses: ingresses,
+		services:  services,
+		watchChan: watchChan,
+	}
+
+	testCases := []struct {
+		desc     string
+		provider Provider
+		expected *types.Configuration
+	}{
+		{
+			desc:     "Empty IngressClass annotation",
+			provider: Provider{},
+			expected: buildConfiguration(
+				backends(
+					backend("other/stuff",
+						servers(
+							server("http://example.com", weight(1)),
+							server("http://example.com", weight(1))),
+						lbMethod("wrr"),
+					),
+					backend("other/",
+						servers(
+							server("http://example.com", weight(1)),
+							server("http://example.com", weight(1))),
+						lbMethod("wrr"),
+					),
+					backend("other/sslstuff",
+						servers(
+							server("http://example.com", weight(1)),
+							server("http://example.com", weight(1))),
+						lbMethod("wrr"),
+					),
+				),
+				frontends(
+					frontend("other/stuff",
+						headers(),
+						passHostHeader(),
+						routes(
+							route("/stuff", "PathPrefix:/stuff"),
+							route("other", "Host:other")),
+					),
+					frontend("other/",
+						headers(),
+						passHostHeader(),
+						routes(
+							route("/", "PathPrefix:/"),
+							route("other", "Host:other")),
+					),
+					frontend("other/sslstuff",
+						headers(),
+						passHostHeader(),
+						passTLSCert(),
+						routes(
+							route("/sslstuff", "PathPrefix:/sslstuff"),
+							route("other", "Host:other")),
+					),
+				),
+			),
+		},
+		{
+			desc:     "Provided IngressClass annotation",
+			provider: Provider{IngressClass: traefikDefaultAnnotationValue + "-other"},
+			expected: buildConfiguration(
+				backends(
+					backend("herp/derp",
+						servers(),
+						lbMethod("wrr"),
+					),
+				),
+				frontends(
+					frontend("herp/derp",
+						headers(),
+						passHostHeader(),
+						routes(
+							route("/derp", "PathPrefix:/derp"),
+							route("herp", "Host:herp")),
+					),
+				),
+			),
+		},
+	}
+
+	for _, test := range testCases {
+		test := test
+		t.Run(test.desc, func(t *testing.T) {
+			t.Parallel()
+
+			actual, err := test.provider.loadIngresses(client)
+			require.NoError(t, err, "error loading ingresses")
+
+			assert.Equal(t, test.expected, actual)
+		})
+	}
+}
```

Comment:
One general note: I'm having major troubles reading this and other Kubernetes tests for several reasons:

- We tend to have a large chunk of (Ingress) fixtures up front and then another chunk of (configuration) expectations near the bottom. I cannot really understand a single test case without either scrolling up and down several times or switching back and forth between two tabs in my browser all the time. The usual Go table-driven style combines both the test fixture and the expectation in a single test case.
- Our test cases are often too coarse and include multiple frontends/backends in the assertion, instead of focusing on a single frontend/backend pair per test case. We should prefer a higher number of relatively small test cases over a smaller number of large-ish test cases.
- We always compare complete configurations. This often requires writing a lot of boilerplate and is inefficient when the kind of functionality we want to validate touches a particular part of the expected configuration only. For instance, in `TestIngressClassAnnotation` above we shouldn't have to worry about host headers, TLS settings, and server weights. Instead, we should only care whether a frontend/backend is included, in the most cheap way conceivable.

Overall, the current structure makes it very difficult to judge if a test is good, complete, and meaningful.

I can understand if we move any major improvement into a follow-up PR; I'd personally prefer though to start having new tests properly structured already.

Reply:
Even if I am agree with you I think it can be done in a new PR :wink:

Reply:
:+1: